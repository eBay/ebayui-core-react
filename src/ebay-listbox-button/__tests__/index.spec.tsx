import initStoryshots from '@storybook/addon-storyshots';
import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { EbayListboxButton, EbayListboxButtonOption } from '..';
jest.useFakeTimers()
describe("<EbayListboxButton>", () => {
    describe("a11y prefix", () => {
        const renderListbox = async (listboxBtnLabel?) => {
            const { getByRole, getByText } = await render(
                <>
                    <label id={listboxBtnLabel}>Select these items</label>
                    <br></br>
                    <br></br>
                    <EbayListboxButton value="BB" prefixId={listboxBtnLabel}>
                        <EbayListboxButtonOption value="AA">Option 1</EbayListboxButtonOption>
                        <EbayListboxButtonOption value="BB">Option 2</EbayListboxButtonOption>
                        <EbayListboxButtonOption value="CC">Option 3</EbayListboxButtonOption>
                    </EbayListboxButton>
                </>
            );
            const buttonElement = getByRole("button");
            const expandBtnTextId = getByText("Option 2").id

            return { buttonElement, expandBtnTextId };
        };

        it("should render with correct aria-labelledby if recived prefix id", async () => {
            const listboxBtnLabel = "listbox-button__label";
            const { buttonElement, expandBtnTextId } = await renderListbox(listboxBtnLabel);

            expect(buttonElement).toHaveAttribute("aria-labelledby", `${listboxBtnLabel} ${expandBtnTextId}`);
        });

        it("should render without aria-labelledby if NOT recived prefix id", async () => {
            const { buttonElement } = await renderListbox();

            expect(buttonElement).not.toHaveAttribute("aria-labelledby");
        });
    });

    describe('given the listbox with 3 items', () => {
        let component
        beforeEach(async () => {
            component = await render(
                    <EbayListboxButton value="BB" prefixId={"listboxBtnLabel"} name="listbox-button-name">
                        <EbayListboxButtonOption value="AA">Option 1</EbayListboxButtonOption>
                        <EbayListboxButtonOption value="BB">Option 2</EbayListboxButtonOption>
                        <EbayListboxButtonOption value="CC">Option 3</EbayListboxButtonOption>
                    </EbayListboxButton>
            );
        });
        it('then it should not be expanded', () => {
            expect(component.getByRole('button')).toHaveAttribute("aria-expanded", `false`);
        });

        describe('when the button is clicked', () => {
            beforeEach(async () => {
                await fireEvent.click(component.getByRole('button'));
            });
            it('then it has expanded the listbox', () => {
                expect(component.getByRole('button')).toHaveAttribute("aria-expanded", `true`);
            });
            it('then listbox options and rendered', () => {
                expect(component.getByRole('listbox')).toBeInTheDocument();
            })
            it('focus should move to listbox', () => {
                const listbox =  component.getByRole('listbox')
                jest.runAllTimers()
                expect(listbox).toHaveFocus();
            })
            describe('when the button is clicked again', () => {
                beforeEach(async () => {
                    await fireEvent.click(component.getByRole('button'));
                });
                it('then it has collapsed the listbox',() => {
                    expect(component.getByRole('button')).toHaveAttribute("aria-expanded", `false`);
                });
                xit('focus should move to button', () => {
                    const button = component.getByRole('button')
                    jest.runAllTimers()
                    expect(button).toHaveFocus();
                })
            });
        });

    })
});

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})
