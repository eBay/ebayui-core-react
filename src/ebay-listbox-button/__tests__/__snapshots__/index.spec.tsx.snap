// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-listbox-button Borderless 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form btn--borderless"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Default - no selected option 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 1
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="AA"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Default - with a11y prefix 1`] = `
Array [
  <label
    id="listbox-button__label"
  >
    Select these items:
  </label>,
  <br />,
  <br />,
  <span
    className="listbox-button"
  >
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="listbox-button__label expand-btn-text"
      className="btn btn--form"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      type="button"
    >
      <span
        className="btn__cell"
      >
        <span
          className="btn__text"
          id="expand-btn-text"
        >
          Option 2
        </span>
        <svg
          aria-hidden={true}
          className="icon icon--dropdown"
          focusable={false}
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlinkHref="#icon-dropdown"
          />
        </svg>
      </span>
    </button>
    <select
      className="listbox-button__native"
      hidden={true}
      value="BB"
    >
      <option
        value="AA"
      />
      <option
        value="BB"
      />
      <option
        value="CC"
      />
    </select>
  </span>,
]
`;

exports[`Storyshots ebay-listbox-button Default 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Disabled State 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form"
    disabled={true}
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Floating label 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form btn--floating-label"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__floating-label"
      >
        Select
      </span>
      <span
        className="btn__text"
      >
        Option 1
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="AA"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Fluid 1`] = `
<span
  className="listbox-button listbox-button--fluid"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Invalid State 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-invalid="true"
    className="btn btn--form"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
  </select>
</span>
`;

exports[`Storyshots ebay-listbox-button Statefull component 1`] = `
<div>
  <span
    className="listbox-button"
  >
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      className="btn btn--form"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      type="button"
    >
      <span
        className="btn__cell"
      >
        <span
          className="btn__text"
        >
          California
        </span>
        <svg
          aria-hidden={true}
          className="icon icon--dropdown"
          focusable={false}
          height="24px"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlinkHref="#icon-dropdown"
          />
        </svg>
      </span>
    </button>
    <select
      className="listbox-button__native"
      hidden={true}
      value="California"
    >
      <option
        value="California"
      />
      <option
        value="New York"
      />
      <option
        value="Washington"
      />
    </select>
  </span>
  <label>
    <input
      name="listbox-value"
      onClick={[Function]}
      type="radio"
    />
    California
  </label>
  <label>
    <input
      name="listbox-value"
      onClick={[Function]}
      type="radio"
    />
    New York
  </label>
  <label>
    <input
      name="listbox-value"
      onClick={[Function]}
      type="radio"
    />
    Washington
  </label>
</div>
`;

exports[`Storyshots ebay-listbox-button Too many options 1`] = `
<span
  className="listbox-button"
>
  <button
    aria-expanded={false}
    aria-haspopup="listbox"
    className="btn btn--form"
    onClick={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    type="button"
  >
    <span
      className="btn__cell"
    >
      <span
        className="btn__text"
      >
        Option 2
      </span>
      <svg
        aria-hidden={true}
        className="icon icon--dropdown"
        focusable={false}
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlinkHref="#icon-dropdown"
        />
      </svg>
    </span>
  </button>
  <select
    className="listbox-button__native"
    hidden={true}
    value="BB"
  >
    <option
      value="AA"
    />
    <option
      value="BB"
    />
    <option
      value="CC"
    />
    <option
      value="DD"
    />
    <option
      value="EE"
    />
    <option
      value="FF"
    />
    <option
      value="GG"
    />
    <option
      value="HH"
    />
    <option
      value="II"
    />
    <option
      value="JJ"
    />
    <option
      value="KK"
    />
    <option
      value="LL"
    />
    <option
      value="MM"
    />
    <option
      value="NN"
    />
    <option
      value="OO"
    />
    <option
      value="PP"
    />
    <option
      value="QQ"
    />
    <option
      value="RR"
    />
    <option
      value="SS"
    />
    <option
      value="TT"
    />
    <option
      value="UU"
    />
    <option
      value="VV"
    />
    <option
      value="YY"
    />
    <option
      value="XX"
    />
    <option
      value="YY"
    />
    <option
      value="ZZ"
    />
    <option
      value="11"
    />
    <option
      value="22"
    />
    <option
      value="33"
    />
    <option
      value="44"
    />
    <option
      value="55"
    />
    <option
      value="66"
    />
    <option
      value="77"
    />
    <option
      value="88"
    />
    <option
      value="99"
    />
    <option
      value="100"
    />
    <option
      value="101"
    />
    <option
      value="102"
    />
    <option
      value="102"
    />
  </select>
</span>
`;
